File: quijote-flutter/lib/processing_engine.dart
class ProcessingResult {
  final int wordCount;
  final int sentenceCount;
  final int paragraphCount;
  final List<MapEntry<String, int>> wordFrequency;
  final double processingTimeMs;

  const ProcessingResult({
    required this.wordCount,
    required this.sentenceCount,
    required this.paragraphCount,
    required this.wordFrequency,
    required this.processingTimeMs,
  });
}

class ProcessingEngine {
  static ProcessingResult process(String text) {
    final stopwatch = Stopwatch()..start();

    // ── Párrafos: contar sin split() ────────────────────────────────────────
    int paragraphCount = 0;
    bool prevBlank = true;
    int lineStart = 0;
    for (int i = 0; i <= text.length; i++) {
      if (i == text.length || text[i] == '\n') {
        final line = text.substring(lineStart, i).trim();
        final isBlank = line.isEmpty;
        if (!isBlank && prevBlank) paragraphCount++;
        prevBlank = isBlank;
        lineStart = i + 1;
      }
    }

    // ── Oraciones: scan de caracteres ───────────────────────────────────────
    int sentenceCount = 0;
    bool inPunct = false;
    for (int i = 0; i < text.length; i++) {
      final c = text[i];
      if (c == '.' || c == '!' || c == '?') {
        if (!inPunct) { sentenceCount++; inPunct = true; }
      } else if (c != ' ' && c != '\n' && c != '\r' && c != '\t') {
        inPunct = false;
      }
    }

    // ── Palabras + frecuencia: un solo paso sin listas intermedias ───────────
    final freqMap = <String, int>{};
    int wordCount = 0;
    int wordStart = -1;

    for (int i = 0; i <= text.length; i++) {
      final isLetter = i < text.length && _isLetter(text.codeUnitAt(i));
      if (isLetter) {
        if (wordStart == -1) wordStart = i;
      } else {
        if (wordStart != -1) {
          final w = text.substring(wordStart, i).toLowerCase();
          freqMap[w] = (freqMap[w] ?? 0) + 1;
          wordCount++;
          wordStart = -1;
        }
      }
    }

    // ── Ordenar por frecuencia ───────────────────────────────────────────────
    final wordFrequency = freqMap.entries.toList()
      ..sort((a, b) => b.value.compareTo(a.value));

    stopwatch.stop();

    return ProcessingResult(
      wordCount: wordCount,
      sentenceCount: sentenceCount,
      paragraphCount: paragraphCount,
      wordFrequency: wordFrequency,
      processingTimeMs: stopwatch.elapsedMicroseconds / 1000.0,
    );
  }

  // Detecta letras Unicode sin regex (cubre latín, español, etc.)
  static bool _isLetter(int codeUnit) {
    // Básico Latin (a-z, A-Z)
    if ((codeUnit >= 65 && codeUnit <= 90) || (codeUnit >= 97 && codeUnit <= 122)) return true;
    // Latín extendido (acentos, ñ, ü, etc.) — U+00C0 a U+024F
    if (codeUnit >= 0x00C0 && codeUnit <= 0x024F) return true;
    // Alfabetos adicionales comunes
    if (codeUnit >= 0x0400 && codeUnit <= 0x04FF) return true; // Cirílico
    if (codeUnit >= 0x0600 && codeUnit <= 0x06FF) return true; // Árabe
    return false;
  }
}

File: quijote-reactnative/src/ProcessingEngine.ts
export interface ProcessingResult {
  wordCount: number;
  sentenceCount: number;
  paragraphCount: number;
  wordFrequency: Array<{ word: string; count: number }>;
  processingTimeMs: number;
}

export function processText(text: string): ProcessingResult {
  const start = performance.now();

  // ── Párrafos: contar sin split() ──────────────────────────────────────────
  let paragraphCount = 0;
  let prevBlank = true;
  let lineStart = 0;
  for (let i = 0; i <= text.length; i++) {
    if (i === text.length || text[i] === '\n') {
      const line = text.slice(lineStart, i).trim();
      const isBlank = line.length === 0;
      if (!isBlank && prevBlank) paragraphCount++;
      prevBlank = isBlank;
      lineStart = i + 1;
    }
  }

  // ── Oraciones: scan de caracteres ─────────────────────────────────────────
  let sentenceCount = 0;
  let inPunct = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (c === '.' || c === '!' || c === '?') {
      if (!inPunct) { sentenceCount++; inPunct = true; }
    } else if (c !== ' ' && c !== '\n' && c !== '\r' && c !== '\t') {
      inPunct = false;
    }
  }

  // ── Palabras + frecuencia: un solo paso sin array intermedio ──────────────
  const freqMap: Record<string, number> = Object.create(null);
  let wordCount = 0;
  let wordStart = -1;
  const LETTER = /\p{L}/u;

  for (let i = 0; i <= text.length; i++) {
    const isLetter = i < text.length && LETTER.test(text[i]);
    if (isLetter) {
      if (wordStart === -1) wordStart = i;
    } else {
      if (wordStart !== -1) {
        const w = text.slice(wordStart, i).toLowerCase();
        freqMap[w] = (freqMap[w] ?? 0) + 1;
        wordCount++;
        wordStart = -1;
      }
    }
  }

  // ── Ordenar por frecuencia ─────────────────────────────────────────────────
  const wordFrequency = Object.entries(freqMap)
    .map(([word, count]) => ({ word, count }))
    .sort((a, b) => b.count - a.count);

  const processingTimeMs = performance.now() - start;

  return { wordCount, sentenceCount, paragraphCount, wordFrequency, processingTimeMs };
}
